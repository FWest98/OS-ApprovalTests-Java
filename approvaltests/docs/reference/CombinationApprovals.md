toc

## Verifying a Variety of Inputs

* [A variety of inputs for a single parameter](/docs/how_to/TestAVarietyOfValues.md)
* [A variety of inputs for multiple parameters](/docs/how_to/TestCombinations.md)
* [A minimum set of inputs for multiple parameters](/docs/Features.md#combinationapprovalsverifybestcoveringpairs)




