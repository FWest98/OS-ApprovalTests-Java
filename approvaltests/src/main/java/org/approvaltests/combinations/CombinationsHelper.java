package org.approvaltests.combinations;

import java.util.ArrayList;
import java.util.List;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.lambda.actions.Action9;
import org.lambda.functions.Function9;

public class CombinationsHelper
{
  public static final Object   EMPTY_ENTRY = new Object();
  public static final Object[] EMPTY       = {EMPTY_ENTRY};
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9, OUT> void verifyAllCombinations(
      Function9<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7,
      IN8[] parameters8, IN9[] parameters9, Options options)
  {
    StringBuffer output = new StringBuffer();
    Action9<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9> foo = (in11, in12, in13, in14, in15, in16, in17, in18,
        in19) -> {
      String result;
      try
      {
        result = "" + call.call(in11, in12, in13, in14, in15, in16, in17, in18, in19);
      }
      catch (SkipCombination e)
      {
        return;
      }
      catch (Throwable t)
      {
        result = String.format("%s: %s", t.getClass().getName(), t.getMessage());
      }
      output.append(
          String.format("%s => %s \n", filterEmpty(in11, in12, in13, in14, in15, in16, in17, in18, in19), result));
    };
    doForAllCombinations(parameters1, parameters2, parameters3, parameters4, parameters5, parameters6, parameters7,
        parameters8, parameters9, foo);
    // TODO: we still end up with 1 file which is not what we wanted in the case of SVGs or other file formats
    Approvals.verify(output, options);
  }
  // TODO: do we want to revert the changes here then again?!
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9> void doForAllCombinations(IN1[] parameters1,
      IN2[] parameters2, IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6,
      IN7[] parameters7, IN8[] parameters8, IN9[] parameters9,
      Action9<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9> action)
  {
    for (IN1 in1 : parameters1)
    {
      for (IN2 in2 : parameters2)
      {
        for (IN3 in3 : parameters3)
        {
          for (IN4 in4 : parameters4)
          {
            for (IN5 in5 : parameters5)
            {
              for (IN6 in6 : parameters6)
              {
                for (IN7 in7 : parameters7)
                {
                  for (IN8 in8 : parameters8)
                  {
                    for (IN9 in9 : parameters9)
                    {
                      action.call(in1, in2, in3, in4, in5, in6, in7, in8, in9);
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  public static List<Object> filterEmpty(Object... objects)
  {
    List<Object> list = new ArrayList<>();
    for (Object object : objects)
    {
      if (object != EMPTY_ENTRY)
      {
        list.add(object);
      }
    }
    return list;
  }
}
