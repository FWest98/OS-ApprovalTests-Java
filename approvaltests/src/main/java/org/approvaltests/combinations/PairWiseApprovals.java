package org.approvaltests.combinations;

import org.approvaltests.combinations.pairwise.PairwiseHelper;
import org.approvaltests.core.Options;
import org.lambda.functions.Function1;
import org.lambda.functions.Function2;
import org.lambda.functions.Function3;
import org.lambda.functions.Function4;
import org.lambda.functions.Function5;
import org.lambda.functions.Function6;
import org.lambda.functions.Function7;
import org.lambda.functions.Function8;
import org.lambda.functions.Function9;

public class PairWiseApprovals
{
  static Object[] EMPTY = CombinationsHelper.EMPTY;
  public static <IN1, OUT> void verifyBestCoveringPairs(Function1<IN1, OUT> call, IN1[] parameters1)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1), parameters1, EMPTY, EMPTY,
        EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, new Options());
  }
  public static <IN1, OUT> void verifyBestCoveringPairs(Function1<IN1, OUT> call, IN1[] parameters1,
      Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1), parameters1, EMPTY, EMPTY,
        EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, options);
  }
  public static <IN1, IN2, OUT> void verifyBestCoveringPairs(Function2<IN1, IN2, OUT> call, IN1[] parameters1,
      IN2[] parameters2)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2), parameters1, parameters2,
        EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, new Options());
  }
  public static <IN1, IN2, OUT> void verifyBestCoveringPairs(Function2<IN1, IN2, OUT> call, IN1[] parameters1,
      IN2[] parameters2, Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2), parameters1, parameters2,
        EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, options);
  }
  public static <IN1, IN2, IN3, OUT> void verifyBestCoveringPairs(Function3<IN1, IN2, IN3, OUT> call,
      IN1[] parameters1, IN2[] parameters2, IN3[] parameters3)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3), parameters1,
        parameters2, parameters3, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, new Options());
  }
  public static <IN1, IN2, IN3, OUT> void verifyBestCoveringPairs(Function3<IN1, IN2, IN3, OUT> call,
      IN1[] parameters1, IN2[] parameters2, IN3[] parameters3, Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3), parameters1,
        parameters2, parameters3, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, options);
  }
  public static <IN1, IN2, IN3, IN4, OUT> void verifyBestCoveringPairs(Function4<IN1, IN2, IN3, IN4, OUT> call,
      IN1[] parameters1, IN2[] parameters2, IN3[] parameters3, IN4[] parameters4)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4), parameters1,
        parameters2, parameters3, parameters4, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, new Options());
  }
  public static <IN1, IN2, IN3, IN4, OUT> void verifyBestCoveringPairs(Function4<IN1, IN2, IN3, IN4, OUT> call,
      IN1[] parameters1, IN2[] parameters2, IN3[] parameters3, IN4[] parameters4, Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4), parameters1,
        parameters2, parameters3, parameters4, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, options);
  }
  public static <IN1, IN2, IN3, IN4, IN5, OUT> void verifyBestCoveringPairs(
      Function5<IN1, IN2, IN3, IN4, IN5, OUT> call, IN1[] parameters1, IN2[] parameters2, IN3[] parameters3,
      IN4[] parameters4, IN5[] parameters5)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5), parameters1,
        parameters2, parameters3, parameters4, parameters5, EMPTY, EMPTY, EMPTY, EMPTY, new Options());
  }
  public static <IN1, IN2, IN3, IN4, IN5, OUT> void verifyBestCoveringPairs(
      Function5<IN1, IN2, IN3, IN4, IN5, OUT> call, IN1[] parameters1, IN2[] parameters2, IN3[] parameters3,
      IN4[] parameters4, IN5[] parameters5, Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5), parameters1,
        parameters2, parameters3, parameters4, parameters5, EMPTY, EMPTY, EMPTY, EMPTY, options);
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, OUT> void verifyBestCoveringPairs(
      Function6<IN1, IN2, IN3, IN4, IN5, IN6, OUT> call, IN1[] parameters1, IN2[] parameters2, IN3[] parameters3,
      IN4[] parameters4, IN5[] parameters5, IN6[] parameters6)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5, n6), parameters1,
        parameters2, parameters3, parameters4, parameters5, parameters6, EMPTY, EMPTY, EMPTY, new Options());
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, OUT> void verifyBestCoveringPairs(
      Function6<IN1, IN2, IN3, IN4, IN5, IN6, OUT> call, IN1[] parameters1, IN2[] parameters2, IN3[] parameters3,
      IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5, n6), parameters1,
        parameters2, parameters3, parameters4, parameters5, parameters6, EMPTY, EMPTY, EMPTY, options);
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, OUT> void verifyBestCoveringPairs(
      Function7<IN1, IN2, IN3, IN4, IN5, IN6, IN7, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5, n6, n7),
        parameters1, parameters2, parameters3, parameters4, parameters5, parameters6, parameters7, EMPTY, EMPTY,
        new Options());
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, OUT> void verifyBestCoveringPairs(
      Function7<IN1, IN2, IN3, IN4, IN5, IN6, IN7, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7,
      Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5, n6, n7),
        parameters1, parameters2, parameters3, parameters4, parameters5, parameters6, parameters7, EMPTY, EMPTY,
        options);
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, OUT> void verifyBestCoveringPairs(
      Function8<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7,
      IN8[] parameters8)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5, n6, n7, n8),
        parameters1, parameters2, parameters3, parameters4, parameters5, parameters6, parameters7, parameters8,
        EMPTY, new Options());
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, OUT> void verifyBestCoveringPairs(
      Function8<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7,
      IN8[] parameters8, Options options)
  {
    verifyBestCoveringPairs((n1, n2, n3, n4, n5, n6, n7, n8, n9) -> call.call(n1, n2, n3, n4, n5, n6, n7, n8),
        parameters1, parameters2, parameters3, parameters4, parameters5, parameters6, parameters7, parameters8,
        EMPTY, options);
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9, OUT> void verifyBestCoveringPairs(
      Function9<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7,
      IN8[] parameters8, IN9[] parameters9)
  {
    verifyBestCoveringPairs(call, parameters1, parameters2, parameters3, parameters4, parameters5, parameters6,
        parameters7, parameters8, parameters9, new Options());
  }
  public static <IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9, OUT> void verifyBestCoveringPairs(
      Function9<IN1, IN2, IN3, IN4, IN5, IN6, IN7, IN8, IN9, OUT> call, IN1[] parameters1, IN2[] parameters2,
      IN3[] parameters3, IN4[] parameters4, IN5[] parameters5, IN6[] parameters6, IN7[] parameters7,
      IN8[] parameters8, IN9[] parameters9, Options options)
  {
    PairwiseHelper.verifyBestCoveringPairs(call, parameters1, parameters2, parameters3, parameters4, parameters5,
        parameters6, parameters7, parameters8, parameters9, options);
  }
}
