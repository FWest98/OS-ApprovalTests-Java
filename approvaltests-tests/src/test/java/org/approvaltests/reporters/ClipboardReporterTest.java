package org.approvaltests.reporters;

import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

class ClipboardReporterTest
{
  @Test
  void commandLineMoves()
  {
    String[] files = {"with spaces.txt", "withoutSpaces.txt"};
    Boolean[] systems = {true, false};
    CombinationApprovals.verifyAllCombinations(ClipboardReporter::getAcceptApprovalText, files, files, systems);
  }
}
