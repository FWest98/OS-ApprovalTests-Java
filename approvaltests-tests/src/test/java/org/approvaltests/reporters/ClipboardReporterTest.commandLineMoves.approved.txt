[with spaces.txt, with spaces.txt, true] => move /Y "with spaces.txt"  "with spaces.txt" 
[with spaces.txt, with spaces.txt, false] => mv "with spaces.txt" "with spaces.txt" 
[with spaces.txt, withoutSpaces.txt, true] => move /Y "with spaces.txt"  "withoutSpaces.txt" 
[with spaces.txt, withoutSpaces.txt, false] => mv "with spaces.txt" withoutSpaces.txt 
[withoutSpaces.txt, with spaces.txt, true] => move /Y "withoutSpaces.txt"  "with spaces.txt" 
[withoutSpaces.txt, with spaces.txt, false] => mv withoutSpaces.txt "with spaces.txt" 
[withoutSpaces.txt, withoutSpaces.txt, true] => move /Y "withoutSpaces.txt"  "withoutSpaces.txt" 
[withoutSpaces.txt, withoutSpaces.txt, false] => mv withoutSpaces.txt withoutSpaces.txt 
